### ISSUE TITLE
Ability to set uniforms by location instead of name, to support multiple SPIR-V entry points

### ISSUE THREAD

**Issue by Boscop** [glium/glium#1904]
Now that glium has support for SPIR-V shaders, here's a new issue that came up:
One SPIR-V module can contain multiple entry points ("main" functions), even for the same shader stage (e.g. fragment shader), which is working fine with glium as long as these entry points don't read the same uniforms!
E.g. if entry points `main_fs` and `main_equi` both read the uniform `iResolution` (let's call this a "shared uniform"), then it will not be one uniform but two separate uniforms with separate `location`!
To investigate this, I modified my local glium fork to print `println!("uniform: {}, location: {}", uniform_name, location);` in `reflect_uniforms` before the line `uniforms.insert(uniform_name, Uniform {..`, then I loaded my 3 entry points in the order in which they appear in the spv module, this is what it prints:
```glsl
--- load entry point: main_fs
uniform: iResolution, location: 1
uniform: internal_vflip, location: 0 // from vertex shader, written in GLSL and separately compiled to SPIR-V
--- load entry point: main_equi
uniform: _equi_mode_52, location: 1
uniform: _fov_rad_54, location: 2
uniform: _stereo_mode_53, location: 3
uniform: _view_57.x_axis, location: 4
uniform: _view_57.y_axis, location: 5
uniform: _view_57.z_axis, location: 6
uniform: iResolution_42, location: 7
uniform: internal_vflip, location: 0 // from vertex shader, written in GLSL and separately compiled to SPIR-V
--- load entry point: main_vr
uniform: internal_vflip, location: 0 // from vertex shader, written in GLSL and separately compiled to SPIR-V
uniform: unCorners[0], location: 1
uniform: unViewport, location: 6
```
Note: In all 3 entry points, the "shared uniforms" are declared the same way, **with the same name**, e.g.:
```rust
#[spirv(descriptor_set = 0)] iResolution: ::spirv_std::storage_class::UniformConstant<::spirv_std::glam::Vec3A>
```
(Note: Without `#[spirv(descriptor_set = 0)]`, they would all get assigned `location` 0, which then causes a linker error at runtime.)

What happens is, these "shared uniforms" get renamed according to the following scheme (presumably by the OpenGL driver):
- The uniform instance of the first entry point gets the original uniform name, e.g. `foo`
- The remaining uniform instances get a suffix appended to their name, of the form `_\d+`, at least on my GPU.

So when the shader gets loaded with glium and uniforms are looked up by name, it always gets the location of that uniform instance that belongs to the first entry point of the module!
This is why the other entry points appear to receive all-zero values for their uniforms, because their uniforms never get set when lookup is done by name!

So, my suggested solution is:
1. Add the ability to set a specific `location` to each uniform via the `spirv` attribute proc-macro (that's an issue for the rust-gpu repo).
2. Add the ability to lookup uniforms by `location` instead of by name, so that their values can be set by location.

---

This suggested solution would also solve the following problem:
When running `spirv-opt` on a spv module, it completely removes the names from all uniforms:
```rust
[INFO ] load entry point: main_fs
uniform: , location: 1
uniform: internal_vflip, location: 0 // from vertex shader, separately compiled
[INFO ] load entry point: main_vr
uniform: , location: 2
uniform: , location: 3
uniform: internal_vflip, location: 0 // from vertex shader, separately compiled
thread 'main' panicked at 'assertion failed: uniform.1.size.is_none()', /home/me/glium/src/program/reflection.rs:304:13
```

If uniform lookup could be done by `location` instead of by name, then this wouldn't be a problem.
So it would solve both of these issues :)

@est31 @willemv What do you think?

**Comment by est31** [comment]
Hmm yeah maybe we should have a `uniform_location` macro or something?

**Comment by willemv** [comment]
I don't think this is the GL driver doing this, but rather the spirv-linker that was responsible creating the merged spirv binary.

For instance, I followed the instructions that @Boscop wrote in the `README.md` of the spirv example.

I.e. I created these two shaders (they're identical, but they're just to demonstrate the example):
* `shader_fs.frag`
```glsl

#version 430

layout(location = 0) uniform vec2 iResolution;

layout(location = 0) out vec4 color;

void main() {
    vec2 q = gl_FragCoord.xy / iResolution;
    color = vec4(q, 1., 1.);
}
```
* `shader_equi.frag`
```glsl

#version 430

layout(location = 0) uniform vec2 iResolution;

layout(location = 0) out vec4 color;

void main() {
    vec2 q = gl_FragCoord.xy / iResolution;
    color = vec4(q, 1., 1.);
}
```

Then I compiled these both seperately, renaming the entry point:
`glslangValidator.exe -G shader_equi.frag --source-entrypoint main -e main_equi -o frag_equi.spv`
`glslangValidator.exe -G shader_fs.frag --source-entrypoint main -e main_fs -o frag_fs.spv`

_Then_ I linked them using `spirv-link`:
`spirv-link.exe frag_equi.spv frag_fs.spv -o frag.spv`

Inspecting the resulting binary with `spirv-dis` shows the two `iResolution` uniforms: one is called just `iResolution` and the other `iResolution_0` :
```
; SPIR-V
; Version: 1.0
; Generator: Khronos; 17
; Bound: 41
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main_equi "main_equi" %gl_FragCoord %color
               OpEntryPoint Fragment %main_fs "main_fs" %gl_FragCoord_0 %color_0
               OpExecutionMode %main_equi OriginLowerLeft
               OpExecutionMode %main_fs OriginLowerLeft
               OpSource GLSL 430
               OpSource GLSL 430
               OpName %main_equi "main_equi"
               OpName %q "q"
               OpName %gl_FragCoord "gl_FragCoord"
               OpName %iResolution "iResolution"
               OpName %color "color"
               OpName %main_fs "main_fs"
               OpName %q_0 "q"
               OpName %gl_FragCoord_0 "gl_FragCoord"
               OpName %iResolution_0 "iResolution"
               OpName %color_0 "color"
               OpDecorate %gl_FragCoord BuiltIn FragCoord
               OpDecorate %iResolution Location 0
               OpDecorate %color Location 0
               OpDecorate %gl_FragCoord_0 BuiltIn FragCoord
               OpDecorate %iResolution_0 Location 0
               OpDecorate %color_0 Location 0
       %void = OpTypeVoid
         %13 = OpTypeFunction %void
      %float = OpTypeFloat 32
    %v2float = OpTypeVector %float 2
%_ptr_Function_v2float = OpTypePointer Function %v2float
    %v4float = OpTypeVector %float 4
%_ptr_Input_v4float = OpTypePointer Input %v4float
%gl_FragCoord = OpVariable %_ptr_Input_v4float Input
%_ptr_UniformConstant_v2float = OpTypePointer UniformConstant %v2float
%iResolution = OpVariable %_ptr_UniformConstant_v2float UniformConstant
%_ptr_Output_v4float = OpTypePointer Output %v4float
      %color = OpVariable %_ptr_Output_v4float Output
    %float_1 = OpConstant %float 1
%gl_FragCoord_0 = OpVariable %_ptr_Input_v4float Input
%iResolution_0 = OpVariable %_ptr_UniformConstant_v2float UniformConstant
    %color_0 = OpVariable %_ptr_Output_v4float Output
  %float_1_0 = OpConstant %float 1
  %main_equi = OpFunction %void None %13
         %23 = OpLabel
          %q = OpVariable %_ptr_Function_v2float Function
         %24 = OpLoad %v4float %gl_FragCoord
         %25 = OpVectorShuffle %v2float %24 %24 0 1
         %26 = OpLoad %v2float %iResolution
         %27 = OpFDiv %v2float %25 %26
               OpStore %q %27
         %28 = OpLoad %v2float %q
         %29 = OpCompositeExtract %float %28 0
         %30 = OpCompositeExtract %float %28 1
         %31 = OpCompositeConstruct %v4float %29 %30 %float_1 %float_1
               OpStore %color %31
               OpReturn
               OpFunctionEnd
    %main_fs = OpFunction %void None %13
         %32 = OpLabel
        %q_0 = OpVariable %_ptr_Function_v2float Function
         %33 = OpLoad %v4float %gl_FragCoord_0
         %34 = OpVectorShuffle %v2float %33 %33 0 1
         %35 = OpLoad %v2float %iResolution_0
         %36 = OpFDiv %v2float %34 %35
               OpStore %q_0 %36
         %37 = OpLoad %v2float %q_0
         %38 = OpCompositeExtract %float %37 0
         %39 = OpCompositeExtract %float %37 1
         %40 = OpCompositeConstruct %v4float %38 %39 %float_1_0 %float_1_0
               OpStore %color_0 %40
               OpReturn
               OpFunctionEnd
```

In other words, I don't think you need to solve this in the code consuming the spirv binary, but rather in the way the spirv binary is produced

**Comment by willemv** [comment]
@Boscop can you inspect the generated SPIR-V file with `spirv-dis` and verify that it indeed contains the two `iResolution` uniforms from the start?

**Comment by Boscop** [comment]
@willemv Yes, I had noticed that as well, that they are separate symbols already in the spv file. But the final uniform name (e.g. `iResolution_42`) is not stored in the spv file and I'm not sure how to predict it. What I meant is that I think that this concrete final name is chosen by the opengl driver (but maybe I'm wrong).
But yes, the best solution would be if it's possible to merge these "multiple instances of the same uniform" already in the linked spv file, such that lookup by name works (because it'll be the same location). Any idea how to merge them during linking?

**Comment by willemv** [comment]
@Boscop I actually don't have any experience with SPIR-V and what the best practices should be for SPIR-V binaries that contain multiple entry points and want to use the same uniforms. Maybe the people from `rust-gpu` have a better idea.

In any case, adding the possibility to bind uniforms by location rather than name in glium could be useful.
